src/libos/src/vm/process_vm.rs

```diff
@@ -440,25 +439,40 @@ impl ProcessVM {
     }
 
     pub fn get_brk(&self) -> usize {
-        self.brk.load(Ordering::SeqCst)
+        *self.brk.read().unwrap()
     }
 
-    pub fn brk(&self, new_brk: usize) -> Result<usize> {
+    pub fn brk(&self, brk: usize) -> Result<usize> {
         let heap_start = self.heap_range.start();
         let heap_end = self.heap_range.end();
 
-        if new_brk >= heap_start && new_brk <= heap_end {
-            self.brk
-                .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |old_brk| Some(new_brk));
-            Ok(new_brk)
+        // Acquire lock first to avoid data-race.
+        let mut brk_guard = self.brk.write().unwrap();
+
+        if brk >= heap_start && brk <= heap_end {
+            // Get page-aligned brk address.
+            let new_brk = align_up(brk, PAGE_SIZE);
+            // Get page-aligned old brk address.
+            let old_brk = align_up(*brk_guard, PAGE_SIZE);
+
+            // Reset the memory when brk shrinks.
+            if new_brk < old_brk {
+                let shrink_brk_range =
+                    VMRange::new(new_brk, old_brk).expect("shrink brk range must be valid");
+                USER_SPACE_VM_MANAGER.reset_memory(shrink_brk_range)?;
```

Comment:
Should brk return `Ok(*brk_guard)` when `reset_memory` returns Error?

Reply:
I think not. If there is an error during `reset_memory`, there must be a fatal error. Just returning `Ok(*brk_guard)` might cover the error.